test/Validate.Layout.cpp

```diff
@@ -0,0 +1,199 @@
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+// Validation tests for Logical Layout
+
+#include "gmock/gmock.h"
+#include "UnitSPIRV.h"
+#include "ValidateFixtures.h"
+
+#include <functional>
+#include <sstream>
+#include <string>
+#include <utility>
+
+using std::function;
+using std::ostream_iterator;
+using std::pair;
+using std::stringstream;
+using std::string;
+using std::tie;
+using std::tuple;
+using std::vector;
+
+using ::testing::HasSubstr;
+
+using ValidateLayout = spvtest::ValidateBase<
+    tuple<int, pair<string, function<bool(int)>>>,
+    SPV_VALIDATE_LAYOUT_BIT>;
+
+namespace {
+
+// returns true if order is equal to VAL
+template <int VAL>
+bool Equals(int order) {
+  return order == VAL;
+};
+
+// returns true if order is between MIN and MAX(inclusive)
+template <int MIN, int MAX>
+bool Range(int order) {
+  return order >= MIN && order <= MAX;
+};
+
+// clang-format off
+// SPIRV source used to test the logical layout
+vector<string> instructions = {
+  "OpCapability Matrix",
+  "OpExtension \"TestExtension\"",
+  "%inst = OpExtInstImport \"GLSL.std.450\"",
+  "OpMemoryModel Logical GLSL450",
+  "OpEntryPoint GLCompute %func \"\"",
+  "OpExecutionMode %func LocalSize 1 1 1",
+  "%str = OpString \"Test String\"",
+  "OpSource GLSL 450 %str \"uniform vec3 var = vec3(4.0);\"",
+  "OpSourceContinued \"void main(){return;}\"",
+  "OpSourceExtension \"Test extension\"",
+  "OpName %id \"MyID\"",
+  "OpMemberName %struct 1 \"my_member\"",
+  "OpDecorate %dgrp RowMajor",
+  "OpMemberDecorate %struct 1 RowMajor",
+  "%dgrp   = OpDecorationGroup",
+  "OpGroupDecorate %dgrp %mat33 %mat44",
+  "%boolt  =  OpTypeBool",
+  "%voidt  =  OpTypeVoid",
+  "%intt   =  OpTypeInt 32 1",
+  "%vec4   =  OpTypeVector %intt 4",
+  "%vec3   =  OpTypeVector %intt 3",
+  "%mat33  =  OpTypeMatrix %vec3 3",
+  "%mat44  =  OpTypeMatrix %vec4 4",
+  "%struct =  OpTypeStruct %intt %mat33",
+  "%vfunct = OpTypeFunction %voidt",
+  "%viifunct =  OpTypeFunction %voidt %intt %intt",
+  // TODO(umar): OpConstant fails because the type is not defined
+  // TODO(umar): OpGroupMemberDecorate
+  "OpLine %str 3 4",
+  "%func   = OpFunction %voidt None %vfunct",
+  "OpFunctionEnd",
+  "%func2   = OpFunction %voidt None %viifunct",
+  "%funcp1 = OpFunctionParameter %intt",
+  "%funcp2 = OpFunctionParameter %intt",
+  "OpReturn",
```

Comment:
Recommend adding an OpLabel, a function-storage-class variable, a load, a store, and maybe one arithmetic operation.
I'm not sure how well that works with the all-combinations nature of the test.